### PR TITLE
Issue #136: Do not add GWT libs (user or dev) if the lib is already a project dependency

### PR DESCRIPTION
**Pull request by mbarkley**
This PR changes the behaviour of the plugin so that dependencies excluded in the plugin and plugin management of the gwt-maven-plugin are not added to the classpath.

### CODE REVIEW

**Comment by mbarkley** [comment]
Hi @ibaca, can you please take a look at this? The build failure seems to not be caused by this commit (I can reproduce it locally on master).

File: src/main/java/org/codehaus/mojo/gwt/AbstractGwtMojo.java [hunk 1]
```diff
@@ -293,6 +321,17 @@ protected boolean isGenerator() {
         return files;
     }
 
+    private List<Exclusion> getProjectPluginExclusions() {
+      final List<Exclusion> exclusions = new ArrayList<Exclusion>();
+      final Plugin plugin = project.getPlugin(Plugin.constructKey(GWT_PLUGIN_GROUP_ID, GWT_PLUGIN_ARTIFACT_ID));
+      for ( final Dependency dep : plugin.getDependencies() )
+      {
+        exclusions.addAll( dep.getExclusions() );
```
Comment (ibaca) [inline src/main/java/org/codehaus/mojo/gwt/AbstractGwtMojo.java]: Sorry, I do not understand the intention of this. Looks like you merge all exclusion from all dependencies, and later on you filter all dependencies by this aggregate exclusions, is this correct? Will be helpful a test case, this also helps to understand the intention 😉. OTOH, this looks a bit hacky, you can always create a custom version with this and deploy to your internal repo. This plugin is totally stable and feature freeze, so maintainability is not a problem.

**Comment by ibaca** [comment]
What do you think about this approach https://github.com/ibaca/mojo.gwt-maven-plugin/commit/25378d990b84cf3d0dee8f3be915902645d2fb0a?

**Comment by csadilek** [comment]
@ibaca thanks for taking a look. The proposed solution allows to use the default maven dependency exclusion mechanism for plugins, for instance in ```pluginManagement```, which in turn allows us to exclude dependencies once (in a root pom) and doesn't require configuring the exclusion in every project separately.

```xml
   <pluginManagement>
     <plugins>
       <plugin>
          <groupId>org.codehaus.mojo</groupId>
          <artifactId>gwt-maven-plugin</artifactId>
          <dependencies>
            <dependency>
              <groupId>com.google.gwt</groupId>
              <artifactId>gwt-dev</artifactId>
              <version>2.8.0-rc1</version>
              <exclusions>
                <exclusion>
                  <groupId>org.eclipse.jetty</groupId>
                  <artifactId>apache-jsp</artifactId>
                </exclusion>         
              </exclusions>
           </dependency>
          </dependencies>
       </plugin>
     </plugins>
    </pluginManagement>
```

gwt-dev in 2.8.0-rc1 doesn't bundle these external dependencies anymore but brings them in transitively via Maven resolution... 

What do you think? We can definitley look into providing a test for this.

**Comment by csadilek** [comment]
As a compromise between the two solutions we could also hide the proposed functionality behind a property (```respectPluginDependencyExclusions```). Then we could still exclude all unneeded and (for us) harmful dependencies in ```pluginManagement```, but it wouldn't pose a risk for anybody else as the property would be false by default. I don't quite see the risk but trust your judgment ;).

Any of the three solutions is better, imo, then deploying a customized version of the plugin temporarily.

File: src/main/java/org/codehaus/mojo/gwt/AbstractGwtMojo.java (hunk 1 above)
Comment (csadilek) [inline src/main/java/org/codehaus/mojo/gwt/AbstractGwtMojo.java]: This is getting just the dependencies of the gwt-maven plugin and collects the exclusions so we don't add them to the classpath later on. Agreed, on the implementation not being pretty, but we couldn't find a better API to do it. Any advice would be much appreciated.

**Comment by ibaca** [comment]
Maybe I am wrong, but I think that your porposal will allow to add exclusion in any dependency ignoring the dependency itself. So for example you can add a dependency to wherever XXX, and inside this random dependency you can just add your exclusions that will apply to the user/dev libs. Am I wrong? This is why I think that you are not actually respecting plugin dependency exclusion, and that why I move the exclusions to the plugin configuration. 
OTOH the plug-in is full of hacking... I'm just the last maintainer... And the plug-in should be released in less than a week, so I'll try to make it less hacky (maybe I'm wrong and it's not hacky), but if I can't I'll merge your pull request.
I thought in other option that consist in create a flag to not to add this user/dev libs, so you need to add it manually in your classpath or in the plugin deps, but might not be trivial.

**Comment by csadilek** [comment]
@ibaca thanks again for your comments!

<Maybe I am wrong, but I think that your porposal will allow to add exclusion in any dependency ignoring the dependency itself>

I've just tested this again and verified that the proposed solution only respects exclusions on the plugin itself, not on any dependency in the project. This works because we collect the dependency exclusions using ```getPlugin().getDependencies()``` which only returns this plugin's dependencies. So, this will only respect exclusions added explicitly to the gwt-maven-plugin.

**Comment by ibaca** [comment]
Yes, I mean the plug-in dependencies... Anyway is almost equivalent, you can add the exclusion in any dependency (yep, plugin dependencies list) and the exclusion will be applied to the internal plugin added libs,  not to the dependency that contains the exclusion.

**Comment by ibaca** [comment]
An example; if you add a GWT plugin dependency to mockito and in the mockito dependency you excludes jetty you get the same result than your example.

**Comment by csadilek** [comment]
Ah, yes. This collects all plugin dependency exclusions, not just for gwt-dev and gwt-user, but the ```getJarFiles``` method is only used for gwt-dev and gwt-user. So, as you write, adding another plugin dependency with exclusions will also affect the classpath (exclusion will be respected/outcome is the same). If that's worth fixing, we can change the logic in ```getProjectPluginExclusions ``` to only add exclusions that come from gwt-user and gwt-dev?

**Comment by ibaca** [comment]
And this one, fixes your problem? https://github.com/ibaca/mojo.gwt-maven-plugin/commit/7d71311de5fa4cf784873a5d3ca26601dfe0a45f

Now the dependency is not added if it's already defined as a project dependency, and if it's defined as project dependencies exclusion should be handled correctly. A bit more aggressive, I hope this does not breaks any project.

**Comment by csadilek** [comment]
I think this should also fix the problem. The disadvantage I see is that it's using ```getDependencies``` and not ```getArtifacts``` so it won't work if gwt-dev is brought in transitively, right? This means we'd have to add gwt-dev (with the exclusions) to every single project. I will do some testing today to confirm.

**Comment by csadilek** [comment]
@ibaca we've confirmed now that your approach works perfectly if you switch from ```getDependencies``` to ```getArtifacts``` in ```getJarFiles```, as then your fix works as expected if gwt-dev is transitively included. As a consequence, it will allow us to declare the exclusions only once and not repeat them in every single project.

The originally proposed fix in this PR is less risky, imo, as it will only affect users with explicit plugin exclusions, which is probably no one at this point.  However, we like your fix better (given the change to ```getArtifacts``` mentioned above) because it involves no extra configuration for us.

Here's the updated version of your commit: https://github.com/csadilek/gwt-maven-plugin/commit/294c77fdf1e5abb6b658545607a5302d0ddbceeb

Please let us know if you want the PR changed to this modified solution or which approach you prefer in general. Thanks again for your help.

**Comment by mbarkley** [comment]
@ibaca As an alternative we have amended the original fix [here](https://github.com/mbarkley/gwt-maven-plugin/commit/341831ff38a48f219f565ca87a68acb8739da1dc) to address the problem of gathering exclusions from dependencies other than gwt-dev and gwt-user.

**Comment by ibaca** [comment]
I prefer https://github.com/csadilek/gwt-maven-plugin/commit/294c77fdf1e5abb6b658545607a5302d0ddbceeb (but I admit that this might be a bit of a personal criteria), if you update the pull request with this commit I'll make some more tests and merge it, or otherwise, if you are sure of your pull request (https://github.com/mbarkley/gwt-maven-plugin/commit/341831ff38a48f219f565ca87a68acb8739da1dc) just merge it yourself (it's actually less risky, I'll grant access) 😉 .

PD: you know that the time dedicated to read and write comments in this issue is half of the time needed to migrate to the new plugin 😬 .

**Comment by mbarkley** [comment]
@ibaca I have updated this PR with your approach (+ @csadilek's change), so it is ready for you to test and merge.

Thanks again for all your help!

**Pull request merged**